Feat/#11 - Feed 유즈케이스 및 컨트롤러 기능 추가

jbh010204 opened the pull request:
> [!WARNING]
> **이슈와 프로젝트(Project) 연결 시 반드시 주의하세요.**
> 이는 업무 효율도를 위해서 사용하는 것이지 때문에 협조 부탁드립니다. 추가로, 완료 된 PR 이라면 리뷰어를 직접적으로 연결해주세요 :)

## 📌 연결된 이슈
- ex) resolve #11 

## 📖 주요 변경 사항
> 변경 된 사항에 대해서 개조식으로 작성해주세요
- Layer별 역류 참조가 일어나지 않도록, DTO를 단방향 형태로 구성
- Feed 상세 조회와 요약 조회를 위해 DTO 를 나눠서 presentation 계층 구성
- Feed 조회 시 DTO 및 다른 도메인의 참조가 필요해 최소한의 참조 형태로만 구성

## 🔖 작업 내용 요약
> 작업한 내용에 대해서, 간단하게 요약해주세요
- Feed 목록 조회, 요약 조회, 상세 조회 기능 구현

## 😎 질문
> 작업하면서 궁금했던 점이나, 공유하고 싶은 내용이 있다면 작성해주세요

## reopen
- #21

jbh010204 committed:
feat : User 및 Community DTO 추가

jbh010204 committed:
feat : User 및 Community DTO 추가

jbh010204 committed:
feat : Feed DTO 및 응답 타입 추가

jbh010204 committed:
feat : Feed DTO 및 응답 타입 추가

jbh010204 committed:
feat : CompanyCommunity repository 추가

jbh010204 committed:
feat : CompanyCommunity repository 추가

jbh010204 committed:
feat : FeedRepository 메서드 추가

jbh010204 committed:
feat : FeedRepository 메서드 추가

jbh010204 committed:
refactor : CompanyCommunity와 Company 연관관계 추가

jbh010204 committed:
refactor : CompanyCommunity와 Company 연관관계 추가

jbh010204 committed:
feat : Feed 조회시 유저 조회를 위한 UserRepository 메서드 추가

jbh010204 committed:
feat : Feed 조회시 유저 조회를 위한 UserRepository 메서드 추가

jbh010204 committed:
feat : Feed 비즈니스 로직 구현

jbh010204 committed:
feat : Feed 비즈니스 로직 구현

jbh010204 committed:
refactor : 피드 서비스 에러 코드로 응답 처리

jbh010204 committed:
refactor : 피드 서비스 에러 코드로 응답 처리

polyglot-k commented on src/main/java/com/example/bak/feed/application/FeedService.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Transactional
+    public FeedResult createFeed(String title, String content, Long communityId, Long userId) {
+        User user = userRepository.findUserById(userId)
+                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));
```
이 부분은 추후에 filter 에서 인가처리로 처리하거나 공통처리로 해도 될 듯해요
지금 수정할 필요는 없고, 인증 인가 부분 고도화 들어가면, 한번 회의해보면 좋을거 같아요

jbh010204 replied:
저도 그렇게 생각합니다. 일단 최소한의 형태로만 잡았어서 회의 진행하면 좋을듯합니다~

polyglot-k commented on src/main/java/com/example/bak/feed/application/dto/CommentInfo.java:
```diff
@@ -0,0 +1,19 @@
+package com.example.bak.feed.application.dto;
+
+import com.example.bak.feed.domain.FeedComment;
+
+/**
+ * FeedComment의 기본 정보를 담는 DTO Feed 상세 조회 시 포함됨
+ */
+public record CommentInfo(
+        Long id,
+        String content
+) {
+
+    public static CommentInfo from(FeedComment comment) {
+        return new CommentInfo(
+                comment.getId(),
+                comment.getComment()
+        );
+    }
+}
```
댓글 부분에서도 작성자가 있으면 좋을 것 같아요

jbh010204 replied:
넵 간단한 형태로만 반영해두겠습니다

polyglot-k commented on src/main/java/com/example/bak/feed/presentation/FeedController.java:
```diff
@@ -0,0 +1,64 @@
+                request.userId()
+        );
+        ApiResponse response = ApiResponseFactory.successVoid("피드를 성공적으로 생성하였습니다.");
+        return ResponseEntity.created(UriUtils.current(feedResult.id())).body(response);
```
created 까지하고, 다음 줄에서 body 하는게 조금 더 가독성이 좋을 것 같아요

jbh010204 replied:
반영하겠습니다!

polyglot-k commented on src/main/java/com/example/bak/feed/domain/FeedRepository.java:
```diff
@@ -1,10 +1,17 @@
 
     List<Feed> findAll();
 
+    Page<Feed> findAll(Pageable pageable);
```
Pageable을 한 이유가 있을까요? 타 부분 대비 어떤 것 때문에 선택했을까요? 😎

jbh010204 replied:
제가 생각하기에 저희 서비스 커뮤니티의 피드 목록을 제공할 때 전체 데이터를 가져오지 않고, 페이지를 나눠서 가져온다고 생각했습니다.
그리고 한 번에 몇 개의 글을 가져올지 정확하게 정하지 않았고 변동이 있을 것 같아, 런타임 시점에서 동적으로 페이지의 수를 다르게 가져올 수 있는 Pageable 이 좋겠다고 생각해서 해당 방향으로 구현하였습니다!

polyglot-k replied:
[단순 궁금] 피드 특성 상 무한 스크롤이 필요로 할수도 있을거 같은데, 무한 스크롤 기반으로 들어간다면, 해당 방식으로 유지해도 괜찮을까요!?

jbh010204 replied:
무한 스크롤 기반이면 커서 기반 페이지네이션으로 바꿔야겠다는 생각이드네요...!

polyglot-k commented on src/main/java/com/example/bak/feed/application/dto/FeedSummary.java:
```diff
@@ -0,0 +1,35 @@
+) {
+
+    public static FeedSummary from(Feed feed) {
+        return new FeedSummary(
+                feed.getId(),
+                feed.getTitle(),
+                UserInfo.from(feed.getAuthor()),
+                CommunityInfo.from(feed.getCommunity()),
+                feed.getComments().size()
+        );
+    }
```
dto 조합형태는 조금 햇갈 릴 수 있어서, 너무 내부에 밀어넣으려는 부분 보다는 메소드 내에서 지역 변수로 받고, 해당 지역 변수를 넣어주는게 가시성이 좋아보이는데 @jbh010204 @BaeJunH0 

다른 인원들은 어떤지 궁금하네요

jbh010204 replied:
저도 그 의견에 동감하는 바 입니다. 너무 한꺼번에 만들려고 해서 조금 욱여넣은 감이 없지않아 있네요 😂

polyglot-k reviewed (approved):
대체적으로 컨벤션을 잘지켜서 수행해주신 것 같습니다. 일단 부분적으로 수정해주시고, merge 진행해주시면 좋을 것 같습니다. 일단 approve 하겠습니다~

jbh010204 committed:
refactor : 지역 변수를 사용하여 DTO 반환 형태 가독성 개선

jbh010204 committed:
refactor : 지역 변수를 사용하여 DTO 반환 형태 가독성 개선

jbh010204 committed:
refactor : Response 반환 형태 가독성 개선

jbh010204 committed:
refactor : Response 반환 형태 가독성 개선

jbh010204 committed:
refactor : Comment 도메인에 작성자 필드 추가

jbh010204 committed:
refactor : Comment 도메인에 작성자 필드 추가

polyglot-k commented on src/main/java/com/example/bak/feed/domain/FeedComment.java:
```diff
@@ -22,20 +22,22 @@ public class FeedComment {
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
 
+    private String author;
```
음..? 이렇게만 추가하면 author 가 들어오나용??

author 는 사실상 user 에 있는건데, 어떤식의 동작을 의도했는지 답변 부탁드려요! @jbh010204

polyglot-k replied:
반정규화 형태일까요? 🫢

jbh010204 replied:
이것도 댓글을 만약 비회원도 달 수 있다고 가정한다면 사실 User 참조를 하지않는게 맞지않나? 라는 생각으로 일단 반환 형태만 표시하긴했습니다...

polyglot-k reviewed (changes requested):
@jbh010204 부분적으로 수정잘하셨는데, 궁금한 부분이 있어서 답변 이후, 확인하고 approve 해드리겠습니다.

polyglot-k committed:
docs : vitepress 문서화 config link 수정

polyglot-k committed:
test : Repository 테스트용 간단한 공통 Stub 객체 추가
- auto_increment 특성 상 id 값의 null 허용
- 데이터의 service usecase 형태를 확인하는 것이 주 포인트

polyglot-k committed:
fix : UserRepository 내에 사용하지 않는 findAll() 메소드 제거

polyglot-k committed:
fix : visibleForTesting 패턴을 통한 id 를 포함한 생성자 추가
- testInstance 라는 메소드를 통해서, 개발 레벨에서의 직접적인 사용을 하지 않도록 추가

polyglot-k committed:
test : AbstractStubRepository 기반 UserRepositoryStub 구현
- Stub 의 특성상 Id 의 생성 주도권은 테스트 코드에 있도록 처리
- Fake 와 Stub 의 형태를 고민했고, 순수 유스케이스의 로직을 테스트하기 위해서 Stub 을 선택

polyglot-k committed:
test : UserServiceUnitTest 유닛 테스트 구현
- Nested class 형태로 테스트 코드 작성
- Repository Stub 으로 주입하여, 실제 DB 와 접근 없이 격리
- Stateless 하게 유지하기 위해서, @BeforeEach 내에 Stub 주입
- createUser, getUserProfile 로직에 대해서 테스트 진행

polyglot-k committed:
chore : BakApplicationTests 제거

polyglot-k committed:
ci : Pull Request를 진행 할 때, Build 및 Test 진행 파이프 라인

polyglot-k committed:
chore : gitignore 추가 (vitepress 관련 불필요한 파일 제외)

polyglot-k committed:
Merge pull request #17 from jobmanri/feat/#16

Feat/#16 - User 도메인 어플리케이션 서비스 UseCase 단위 테스트 작성

jbh010204 committed:
refactor : FeedComment User 참조 형태로 변경

jbh010204 committed:
refactor : FeedComment User 참조 형태로 변경

jbh010204 committed:
feat: Comment 기능을 위한 DTO 생성

jbh010204 committed:
feat: Comment 기능을 위한 DTO 생성

jbh010204 committed:
refactor: CommentInfo DTO 수정

jbh010204 committed:
refactor: CommentInfo DTO 수정

jbh010204 committed:
feat: Comment Repository 생성

jbh010204 committed:
feat: Comment Repository 생성

jbh010204 committed:
feat: Comment 로직 및 컨트롤러 구현

jbh010204 committed:
feat: Comment 로직 및 컨트롤러 구현

jbh010204 committed:
Merge remote-tracking branch 'origin/Feat/#11' into Feat/#11

polyglot-k commented:
- #21 로 이동
